packages/text-input/src/TextInput.tsx

```diff
@@ -0,0 +1,331 @@
+ * @param props.setValue Callback used to set the value of the input field.
+ * @param props.className className supplied to the TextInput container.
+ */
+const TextInput = React.forwardRef(
+  (
+    {
+      label,
+      description,
+      optional = false,
+      disabled = false,
+      onChange,
+      placeholder,
+      errorMessage,
+      state,
+      value: controlledValue,
+      setValue: setControlledValue,
+      className,
+    }: TextInputProps,
+    forwardRef: React.Ref<HTMLInputElement>,
+  ) => {
+    const isControlled = typeof controlledValue === 'string';
+    const [uncontrolledValue, setUncontrolledValue] = useState('');
+    const value = isControlled ? controlledValue : uncontrolledValue;
+    const setValue = isControlled ? setControlledValue : setUncontrolledValue;
+    const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+    const [hasFocus, setHasFocus] = useState(false);
+    const inputSelectorProp = createDataProp('input-selector');
+
+    function getInputColorFromState() {
+      if (state === State.None) {
+        return '#CCC';
+      }
+
+      return state === State.Valid ? uiColors.green.base : uiColors.red.base;
+    }
+
+    function getInputPaddingFromState() {
+      if (state === State.Valid || state === State.Error) {
+        return '30px';
+      }
+
+      return optional ? '60px' : '12px';
+    }
+
+    function onValueChange(e: React.ChangeEvent<HTMLInputElement>) {
+      if (onChange !== undefined) {
+        state = onChange(e);
+      }
+
+      if (setValue != undefined) {
+        setValue(e.target.value);
+      }
+    }
+
+    const interactionRingFocusStyle = css`
+      ${inputSelectorProp.selector}:focus ~ & {
+        background-color: #9dd0e7;
+        transform: scale(1);
+        z-index: 1;
+      }
+    `;
+
+    const interactionRingHoverStyle = css`
+      ${inputSelectorProp.selector}:hover ~ & {
+        transform: scale(1);
+      }
+    `;
+
+    const descriptionStyle = css`
+      color: ${uiColors.gray.dark1};
+      font-size: 14px;
+      height: ${description === '' ? '0' : '20'}px;
+      font-weight: normal;
+      padding-bottom: 8px;
+    `;
+
+    const conditionalStyling = css`
+      z-index: ${hasFocus ? 2 : 1};
+      background: ${disabled ? uiColors.gray.light2 : '#FFFFFF'};
+      padding-right: ${getInputPaddingFromState()};
+      border: 1px solid ${getInputColorFromState()};
+    `;
+
+    return (
+      <label className={cx(textInputStyle, labelStyle, className)}>
+        {label}
+        <label className={descriptionStyle}>{description}</label>
+        <div className={inputContainerStyle}>
+          <input
+            {...inputSelectorProp.prop}
+            className={cx(inputStyle, conditionalStyling)}
+            type="text"
+            value={value}
+            required={!optional}
+            disabled={disabled}
+            placeholder={placeholder}
+            onChange={e => onValueChange(e)}
```

Comment:
FYI this is also valid. Rather than making another inline function you can just pass the function reference directly. 

```suggestion
            onChange={onValueChange}
```